### PR TITLE
inspect stopped instances

### PR DESCRIPTION
**Pull request by nahum-litvin-hs**
add inspection for stopped tasks. breaking deployment when tasks are stopped. this is a scenario I encounter very often. 
'ecs deploy' doesn't handle errors like 'EssentialContainerExited ' and we are forced to go to the UI to know why deployment failed,

after this change deploying a bad task with a bad container will end by:
```

Updating service
Successfully changed task definition to: xxxxxxxx:159

Deploying new task definition................................................................................................
2021-06-16 10:55:44.967000
ERROR: EssentialContainerExited

Deployment failed
```

### CODE REVIEW

File: ecs_deploy/cli.py [hunk 1]
```diff
@@ -28,53 +28,96 @@ def get_client(access_key_id, secret_access_key, region, profile):
...
+              help='Overwrites the privileged value for a container: <container> <privileged>')
+@click.option('--essential', type=(str, bool), multiple=True,
+              help='Overwrites the essential value for a container: <container> <essential>')
+@click.option('-e', '--env', type=(str, str, str), multiple=True,
+              help='Adds or changes an environment variable: <container> <name> <value>')
+@click.option('--env-file', type=(str, str), default=((None, None),), multiple=True, required=False,
+              help='Load environment variables from .env-file')
+@click.option('-s', '--secret', type=(str, str, str), multiple=True,
+              help='Adds or changes a secret environment variable from the AWS Parameter Store (Not available for Fargate): <container> <name> <parameter name>')
+@click.option('-u', '--ulimit', type=(str, str, int, int), multiple=True,
```
Comment (fabfuel) [inline ecs_deploy/cli.py]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

It appears you are unpacking more than three variables while using a function that returns more than one variable. Using a large number of return values is prone to errors. We recommend that you return a small class or `namedtuple` instance instead.

[Learn more](https://docs.python.org/3/library/typing.html#typing.NamedTuple)

File: ecs_deploy/cli.py [hunk 2]
```diff
@@ -163,46 +207,79 @@ def deploy(cluster, service, tag, image, command, health_check, cpu, memory, mem
...
+              help='Adds or changes a ulimit variable in the container description (Not available for Fargate): <container> <ulimit name> <softlimit value> <hardlimit value>')
+@click.option('--system-control', type=(str, str, str), multiple=True,
+              help='Adds or changes a system control variable in the container description (Not available for Fargate): <container> <namespace> <value>')
+@click.option('-p', '--port', type=(str, int, int), multiple=True,
+              help='Adds or changes a port mappings in the container description (Not available for Fargate): <container> <container port value> <host port value>')
+@click.option('-m', '--mount', type=(str, str, str), multiple=True,
+              help='Adds or changes a mount points in the container description (Not available for Fargate): <container> <container port value> <host port value>')
+@click.option('-l', '--log', type=(str, str, str, str), multiple=True,
+              help='Adds or changes a log configuration in the container description (Not available for Fargate): <container> <log driver> <option name> <option value>')
```
Comment (fabfuel) [inline ecs_deploy/cli.py]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

It appears you are unpacking more than three variables while using a function that returns more than one variable. Using a large number of return values is prone to errors. We recommend that you return a small class or `namedtuple` instance instead.

[Learn more](https://docs.python.org/3/library/typing.html#typing.NamedTuple)

File: ecs_deploy/cli.py [hunk 3]
```diff
@@ -163,46 +207,79 @@ def deploy(cluster, service, tag, image, command, health_check, cpu, memory, mem
...
+              help='Set the given ulimits exclusively and remove all other pre-existing ulimits from all containers')
+@click.option('--exclusive-system-controls', is_flag=True, default=False,
+              help='Set the given system controls exclusively and remove all other pre-existing system controls from all containers')
+@click.option('--exclusive-ports', is_flag=True, default=False,
+              help='Set the given port mappings exclusively and remove all other pre-existing port mappings from all containers')
+@click.option('--exclusive-mounts', is_flag=True, default=False,
+              help='Set the given mount points exclusively and remove all other pre-existing mount points from all containers')
+@click.option('--volume', type=(str, str), multiple=True, required=False,
+              help='Set volume mapping from host to container in the task definition.')
+def cron(cluster, task, rule, image, tag, command, cpu, memory, memoryreservation, privileged, env, env_file, secret,
```
Comment (fabfuel) [inline ecs_deploy/cli.py]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Using more than five non-keyword parameters in a function definition can indicate that the function is difficult to understand and maintain. Try to combine parameters in objects or refactor the function into small functions.

Comment (fabfuel) [inline ecs_deploy/cli.py]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This function contains 99 lines of code, not including blank lines or lines with only comments, Python punctuation characters, identifiers, or literals. By comparison, 99% of the functions in the CodeGuru reference dataset contain fewer lines of code. Large functions might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this function or break it into multiple functions.

File: ecs_deploy/cli.py [hunk 4]
```diff
@@ -163,46 +207,79 @@ def deploy(cluster, service, tag, image, command, health_check, cpu, memory, mem
...
+@click.option('--memoryreservation', type=(str, int), multiple=True,
+              help='Overwrites the memory reservation value for a container: <container> <memoryreservation>')
+@click.option('--privileged', type=(str, bool), multiple=True,
+              help='Overwrites the memory reservation value for a container: <container> <memoryreservation>')
+@click.option('-e', '--env', type=(str, str, str), multiple=True,
+              help='Adds or changes an environment variable: <container> <name> <value>')
+@click.option('-s', '--secret', type=(str, str, str), multiple=True,
+              help='Adds or changes a secret environment variable from the AWS Parameter Store (Not available for Fargate): <container> <name> <parameter name>')
+@click.option('-u', '--ulimit', type=(str, str, int, int), multiple=True,
+              help='Adds or changes a ulimit variable in the container description (Not available for Fargate): <container> <ulimit name> <softlimit value> <hardlimit value>')
```
Comment (fabfuel) [inline ecs_deploy/cli.py]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

It appears you are unpacking more than three variables while using a function that returns more than one variable. Using a large number of return values is prone to errors. We recommend that you return a small class or `namedtuple` instance instead.

[Learn more](https://docs.python.org/3/library/typing.html#typing.NamedTuple)

File: ecs_deploy/cli.py [hunk 5]
```diff
@@ -28,53 +28,96 @@ def get_client(access_key_id, secret_access_key, region, profile):
...
+@click.option('--system-control', type=(str, str, str), multiple=True,
+              help='Adds or changes a system control variable in the container description (Not available for Fargate): <container> <namespace> <value>')
+@click.option('-p', '--port', type=(str, int, int), multiple=True,
+              help='Adds or changes a port mappings in the container description (Not available for Fargate): <container> <container port value> <host port value>')
+@click.option('-m', '--mount', type=(str, str, str), multiple=True,
+              help='Adds or changes a mount points in the container description (Not available for Fargate): <container> <container port value> <host port value>')
+@click.option('-l', '--log', type=(str, str, str, str), multiple=True,
```
Comment (fabfuel) [inline ecs_deploy/cli.py]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

It appears you are unpacking more than three variables while using a function that returns more than one variable. Using a large number of return values is prone to errors. We recommend that you return a small class or `namedtuple` instance instead.

[Learn more](https://docs.python.org/3/library/typing.html#typing.NamedTuple)

File: ecs_deploy/cli.py [hunk 6]
```diff
@@ -324,10 +410,14 @@ def update(task, image, tag, command, env, env_file, secret, role, region, acces
...
-@click.option('--ignore-warnings', is_flag=True, help='Do not fail deployment on warnings (port already in use or insufficient memory/CPU)')
-@click.option('--sleep-time', default=1, type=int, help='Amount of seconds to wait between each check of the service (default: 1)')
-def scale(cluster, service, desired_count, access_key_id, secret_access_key, region, profile, timeout, ignore_warnings, sleep_time):
+@click.option('--timeout', default=300, type=int,
+              help='Amount of seconds to wait for deployment before command fails (default: 300). To disable timeout (fire and forget) set to -1')
+@click.option('--ignore-warnings', is_flag=True,
+              help='Do not fail deployment on warnings (port already in use or insufficient memory/CPU)')
+@click.option('--sleep-time', default=1, type=int,
+              help='Amount of seconds to wait between each check of the service (default: 1)')
+def scale(cluster, service, desired_count, access_key_id, secret_access_key, region, profile, timeout, ignore_warnings,
```
Comment (fabfuel) [inline ecs_deploy/cli.py]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Using more than five non-keyword parameters in a function definition can indicate that the function is difficult to understand and maintain. Try to combine parameters in objects or refactor the function into small functions.

File: ecs_deploy/cli.py [hunk 7]
```diff
@@ -364,22 +454,34 @@ def scale(cluster, service, desired_count, access_key_id, secret_access_key, reg
...
 @click.option('--region', help='AWS region (e.g. eu-central-1)')
 @click.option('--access-key-id', help='AWS access key id')
 @click.option('--secret-access-key', help='AWS secret access key')
 @click.option('--profile', help='AWS configuration profile name')
-@click.option('--exclusive-env', is_flag=True, default=False, help='Set the given environment variables exclusively and remove all other pre-existing env variables from all containers')
+@click.option('--exclusive-env', is_flag=True, default=False,
+              help='Set the given environment variables exclusively and remove all other pre-existing env variables from all containers')
 @click.option('--diff/--no-diff', default=True, help='Print what values were changed in the task definition')
-def run(cluster, task, count, command, env, env_file, secret, launchtype, subnet, securitygroup, public_ip, platform_version, region, access_key_id, secret_access_key, profile, exclusive_env, diff):
+def run(cluster, task, count, command, env, env_file, secret, launchtype, subnet, securitygroup, public_ip,
```
Comment (fabfuel) [inline ecs_deploy/cli.py]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Using more than five non-keyword parameters in a function definition can indicate that the function is difficult to understand and maintain. Try to combine parameters in objects or refactor the function into small functions.

File: ecs_deploy/cli.py [hunk 8]
```diff
@@ -28,53 +28,96 @@ def get_client(access_key_id, secret_access_key, region, profile):
...
+              help='Set the given port mappings exclusively and remove all other pre-existing port mappings from all containers')
+@click.option('--exclusive-mounts', is_flag=True, default=False,
+              help='Set the given mount points exclusively and remove all other pre-existing mount points from all containers')
+@click.option('--volume', type=(str, str), multiple=True, required=False,
+              help='Set volume mapping from host to container in the task definition.')
+@click.option('--add-container', type=str, multiple=True, required=False,
+              help='Add a placeholder container in the task definition.')
+@click.option('--remove-container', type=str, multiple=True, required=False,
+              help='Remove a container from the task definition.')
+def deploy(cluster, service, tag, image, command, health_check, cpu, memory, memoryreservation, privileged, essential,
```
Comment (fabfuel) [inline ecs_deploy/cli.py]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This function contains 116 lines of code, not including blank lines or lines with only comments, Python punctuation characters, identifiers, or literals. By comparison, 99% of the functions in the CodeGuru reference dataset contain fewer lines of code. Large functions might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this function or break it into multiple functions.

Comment (fabfuel) [inline ecs_deploy/cli.py]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Using more than five non-keyword parameters in a function definition can indicate that the function is difficult to understand and maintain. Try to combine parameters in objects or refactor the function into small functions.

**Comment by fabfuel** [comment]
Hi @nahum-litvin-hs, 

thanks for your PR!
Please add a description, of what the change aims to do and also tests would be needed to merge it.

Best
Fabian

**Comment by nahum-litvin-hs** [comment]
Hi @fabfuel 
Im a little overwhelmed. would you kindly take a first look and provide a more specific advice how to proceed?

**Comment by fabfuel** [comment]
Hi @nahum-litvin-hs,

thanks for adding the descrption!

Would you do me a favor? In the `develop` branch we have some changes about how ecs-deploy checks, if a deployment has succeeded or failed. Have you tried your workload already with the unreaeased `develop` version of the tool? It might be, that this already solves the issue you described in the description, as ECS now keeps track of the deployment success itself and we can utilize this in ecs-deploy, too.

Best
Fabian

**Comment by fabfuel** [comment]
Depending on how you install and use ecs-deploy, running it via Docker (see https://github.com/fabfuel/ecs-deploy#run-via-docker) would be straight forward as:
```
docker run fabfuel/ecs-deploy:develop
```

**Comment by nahum-litvin-hs** [comment]
@fabfuel I have started from develop branch on develop
`ecs-deploy/ecs_deploy/cli.py deploy cluster_name service_name --image container fabfuel/ecs-deploy`
runs to timeout
```Deploying new task definition................................................................................................................warning: 1 tasks failed to start
.................................................................................................```


if you try same on my branch

```Deploying new task definition..............................................................................................
2021-06-16 13:00:00.302000
ERROR: EssentialContainerExited

Deployment failed```

**Comment by fabfuel** [comment]
Did you enable the deployment circuit breaker for your service? Then it still times out? Please see:
- https://aws.amazon.com/blogs/containers/announcing-amazon-ecs-deployment-circuit-breaker/
- https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_DeploymentCircuitBreaker.html

The changes in the `develop` branch should exactly behave as in your example. If the service's tasks fail, the service's rollout state should change to `FAILED` and ecs-deploy shows the deployment had failed.

**Comment by nahum-litvin-hs** [comment]
> Did you enable the deployment circuit breaker for your service? Then it still times out? Please see:
> 
> * https://aws.amazon.com/blogs/containers/announcing-amazon-ecs-deployment-circuit-breaker/
> * https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_DeploymentCircuitBreaker.html
> 
> The changes in the `develop` branch should exactly behave as in your example. If the service's tasks fail, the service's rollout state should change to `FAILED` and ecs-deploy shows the deployment had failed.

see https://github.com/aws/containers-roadmap/issues/1206

**Comment by fabfuel** [comment]
Oh, this is still open. I referred to the same issue back in February, but did not check on it since then. 

Then I think it makes sense to check for the failed containers in the client. But I would move this logic to the `ecs` module, not the `cli`, as there is already all the other logic located around checking for the deployment success or failure: https://github.com/fabfuel/ecs-deploy/blob/develop/ecs_deploy/ecs.py#L1163-L1182

**Pull request closed without merging**